src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,443 @@
+            ExpandOrCollapse(this.DocumentSymbolViewModelItems, shouldExpand);
+
+            static void ExpandOrCollapse(ImmutableArray<DocumentSymbolDataViewModel> models, bool shouldExpand)
+            {
+                foreach (var model in models)
+                {
+                    model.IsExpanded = shouldExpand;
+                    ExpandOrCollapse(model.Children, shouldExpand);
+                }
+            }
+        }
+
+        private async ValueTask ComputeViewStateAsync(CancellationToken cancellationToken)
+        {
+            // We do not want this work running on a background thread
+            await TaskScheduler.Default;
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Do the expensive LSP work of actually computing the items to show.
+            var (documentSymbolData, newTextSnapshot) = await ComputeDocumentSymbolDataAsync(cancellationToken).ConfigureAwait(false);
+
+            // Now, go back to the UI and grab the prior view state we set, and the current UI values we want to update the data with.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var searchText = this.SearchText;
+            var sortOption = this.SortOption;
+            var lastPresentedViewState = this.LastPresentedViewState;
+
+            // Jump back to the BG to do all our work.
+            await TaskScheduler.Default;
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var searchTextChanged = searchText != lastPresentedViewState.SearchText;
+            var oldViewModelItems = lastPresentedViewState.ViewModelItems;
+
+            // if we got new data or the user changed the search text, recompute our items to correspond to this new state.
+            // Apply whatever the current search text is to what the model returned, and produce the new items.
+            var newViewModelItems = GetDocumentSymbolItemViewModels(
+                sortOption, SearchDocumentSymbolData(documentSymbolData, searchText, cancellationToken));
+
+            // If the search text changed, just show everything in expanded form, so the user can see everything
+            // that matched, without anything being hidden.
+            //
+            // in the case of no search text change, attempt to keep the same open/close expansion state from before.
+            if (!searchTextChanged)
+            {
+                ApplyOldStateToNewItems(
+                    oldSnapshot: lastPresentedViewState.TextSnapshot,
+                    newSnapshot: newTextSnapshot,
+                    oldItems: oldViewModelItems,
+                    newItems: newViewModelItems);
+            }
+
+            // Now create an interval tree out of the view models.  This will allow us to easily find the intersecting
+            // view models given any position in the file with any particular text snapshot.
```

Comment:
Was the prior code expensive enough that finding this was a problem? Otherwise we're spending costs here building the tree, is that just going to make more allocations than needed?

Reply:
the prior code was linear... which def worries me in terms of how it will scale up.  It also means we have to be much more careful about hte UI thread (to the point that i would do the computation off of it).  This approach makes it effectively free (At the cost of those allocations).  I'm not terribly worries here though as we use interval trees for all our taggers (including classification), and the memory is fine.  THe number of nodes in this tree are just however many top level symbols there are, so i think the allocation cost here is a fine tradeoff.  

--

another way to think about this: i view this entire feature as just another tagger, with a specialized UI.  so following the same patterns/practices here seems totally ok.